### PR TITLE
Pscanrules: Fixed polyfill rule running slowly

### PR DESCRIPTION
**Pull request by psiinon**
## Overview
The polyfill rule can run very slowly in some cases.
The culpret was the code  which searched for matching URLs in inline scripts.
The fix is to search for all URLs first, and then do a "contains" conditional on the domain name.
The makes a huge difference in examples that previously ran very slowly.

## Checklist
- [ ] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by psiinon** [comment]
Tests now passing thanks to @thc202 😁

**Review by kingthorin — APPROVED** [review]

File: addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java [hunk 1]
```diff
@@ -45,26 +45,41 @@ public class PolyfillCdnScriptScanRule extends PluginPassiveScanner
...
-            Pattern.compile("http[s]?://.*unionadjs\\.com/.*", Pattern.CASE_INSENSITIVE);
-    private static Pattern XHSBPZA_COM =
-            Pattern.compile("http[s]?://.*xhsbpza\\.com/.*", Pattern.CASE_INSENSITIVE);
-    private static Pattern UNION_MACOMS_LA =
-            Pattern.compile("http[s]?://.*union\\.macoms\\.la/.*", Pattern.CASE_INSENSITIVE);
-    private static Pattern NEWCRBPC_COM =
-            Pattern.compile("http[s]?://.*newcrbpc\\.com/.*", Pattern.CASE_INSENSITIVE);
-
-    private static Pattern[] ALL_DOMAINS = {
+    private static String START_P = "http[s]?://.*";
```
Comment (thc202) [inline addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java:48]: These all could be final.

File: addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java [hunk 2]
```diff
@@ -101,12 +128,22 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
                 // Check the script contents, in case they are loading scripts via JS
                 for (Element sourceElement : sourceElements) {
                     String contents = sourceElement.getContent().toString();
-
-                    for (Pattern pattern : ALL_DOMAINS) {
-                        Matcher matcher = pattern.matcher(contents);
-                        if (matcher.find()) {
-                            this.createLowConfidenceAlert(null, matcher.group(0)).raise();
-                            break;
+                    String contentsLc = contents.toLowerCase();
```
Comment (thc202) [inline addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java:131]: `Locale.ROOT`

File: addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java [hunk 3]
```diff
@@ -101,12 +128,22 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
...
+
+                    Matcher urlMatcher = ALL_URLS.matcher(contents);
```
Comment (thc202) [inline addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java:133]: Did you try without matching first for URLs? That seems to be equally fast.

Comment (psiinon) [reply addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/PolyfillCdnScriptScanRule.java]: Tried it and yes, it is as fast 😁 Reverted to not extract the URLs as it makes the code simpler.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**